libyul/Object.cpp

```diff
@@ -61,13 +61,48 @@ string Object::toString(Dialect const* _dialect) const
 	return "object \"" + name.str() + "\" {\n" + indent(inner) + "\n}";
 }
 
-set<YulString> Object::dataNames() const
+set<YulString> Object::qualifiedDataNames() const
 {
-	set<YulString> names;
-	names.insert(name);
-	for (auto const& subObject: subIndexByName)
-		names.insert(subObject.first);
+	set<YulString> qualifiedNames = {name};
+	for (shared_ptr<ObjectNode> const& subObjectNode: subObjects)
+	{
+		qualifiedNames.insert(subObjectNode->name);
+		if (auto const* subObject = dynamic_cast<Object const*>(subObjectNode.get()))
+			for (YulString const& subSubObj: subObject->qualifiedDataNames())
+				qualifiedNames.insert(YulString{subObject->name.str() + "." + subSubObj.str()});
+	}
+
 	// The empty name is not valid
-	names.erase(YulString{});
-	return names;
+	qualifiedNames.erase(YulString{});
+
+	return qualifiedNames;
+}
+
+tuple<bool, vector<size_t>> Object::pathToSubObject(string _qualifiedName) const
+{
+	if (_qualifiedName == name.str())
+		return {};
+	if (_qualifiedName.rfind(name.str() + ".", 0) == 0)
+		_qualifiedName.erase(0, name.str().length() + 1);
```

Comment:
Maybe use substr instead?

Reply:
> The correctness of this removal strongly depends on this being true:
> 
> ```c++
> yulAssert(subObjects.count(name.str()) == 0, "");
> ```
> 
> I think you should be explicit about it.

I am not sure if I follow. This assert would mean that object `A` can have subobject with same name (`A`).
I don't mind adding assert, but I think it will never be activated.

Will use substr instead of erasing

Reply:
I meant that you can only ignore the prefix in `_qualifiedName` if you're sure that `subObjects` cannot contain the name of the current object. And that's a correct assumption but it's enforced elsewhere and it would be good to have a sanity check here has well. If it ever stops being true, this code will break.

Reply:
Added a check